naga/src/compact/mod.rs

```diff
@@ -880,3 +934,75 @@ fn local_expression_override() {
     compact(&mut module);
     assert!(validator.validate(&module).is_ok());
 }
+
+#[test]
+fn unnamed_constant_type() {
+    let mut module = crate::Module::default();
+    let nowhere = crate::Span::default();
+
+    // This type is used only by the unnamed constant.
+    let ty_u32 = module.types.insert(
+        crate::Type {
+            name: None,
+            inner: crate::TypeInner::Scalar(crate::Scalar::U32),
+        },
+        nowhere,
+    );
+
+    // This type is used by the named constant.
+    let ty_vec_u32 = module.types.insert(
+        crate::Type {
+            name: None,
+            inner: crate::TypeInner::Vector {
+                size: crate::VectorSize::Bi,
+                scalar: crate::Scalar::U32,
+            },
+        },
+        nowhere,
+    );
+
+    let unnamed_init = module
+        .global_expressions
+        .append(crate::Expression::Literal(crate::Literal::U32(0)), nowhere);
+
+    let unnamed_constant = module.constants.append(
+        crate::Constant {
+            name: None,
+            ty: ty_u32,
+            init: unnamed_init,
+        },
+        nowhere,
+    );
+
+    // The named constant is initialized using a Splat expression, to
+    // give the named constant a type distinct from the unnamed
+    // constant's.
+    let unnamed_constant_expr = module
+        .global_expressions
+        .append(crate::Expression::Constant(unnamed_constant), nowhere);
+    let named_init = module.global_expressions.append(
+        crate::Expression::Splat {
+            size: crate::VectorSize::Bi,
+            value: unnamed_constant_expr,
+        },
+        nowhere,
+    );
+
+    let _named_constant = module.constants.append(
+        crate::Constant {
+            name: Some("totally_named".to_string()),
+            ty: ty_vec_u32,
+            init: named_init,
+        },
+        nowhere,
+    );
+
+    let mut validator = super::valid::Validator::new(
+        super::valid::ValidationFlags::all(),
+        super::valid::Capabilities::all(),
+    );
+
+    assert!(validator.validate(&module).is_ok());
+    compact(&mut module);
+    assert!(validator.validate(&module).is_ok());
```

Comment:
nitpick: For this and the other test being added: Checking that the module still passes validation afterward is good, but I'm surprised that we're not snapshotting the contents of the module's arenas so we can enforce that certain things gets culled, and others don't, given the context of this PR. After all, the commits adding these tests are plugging a case where we were accidentally removing things, right?

Reply:
The bug being addressed in this PR is that compaction was removing stuff that was being used, leaving dangling handles behind. So while it's true that we do know exactly what we expect the module to contain after compaction, meaning that a snapshot test wouldn't be wrong, checking validity should catch the failure reliably.

We agreed in chat that this wasn't a blocking concern.